addons/hw_drivers/tools/route.py

```diff
@@ -0,0 +1,37 @@
+import functools
+import logging
+
+from odoo.addons.hw_drivers.tools import helpers
+from werkzeug.exceptions import Forbidden, BadRequest
+
+_logger = logging.getLogger(__name__)
+
+
+def protect(endpoint):
+    """Decorate a route to protect it with TOTP. If the IoT Box is not connected
+    to a db, the route will not be protected.
+    """
+    fname = f"<function {endpoint.__module__}.{endpoint.__qualname__}>"
+
+    @functools.wraps(endpoint)
+    def protect_wrapper(*args, token=None, **kwargs):
+        # If no db connected, we don't protect the endpoint
+        if not helpers.get_odoo_server_url():
+            return endpoint(*args, **kwargs)
+
+        # Check if token is provided
+        if not token:
+            _logger.error('%s: Authentication required.', fname)
+            return Forbidden('Authentication required.')
+
+        # Check token type
+        if not isinstance(token, str):
+            _logger.error('%s, Token should be a string', fname)
+            return BadRequest('Invalid token format.')
+
+        if token != helpers.get_token():
```

Comment:
```suggestion
        if odoo.tools.consteq(token, helpers.get_token()):
```

In addition to XMO's review, here an improvement you should add.
Testing the token's validity using `consteq` avoid [timing attacks](https://en.wikipedia.org/wiki/Timing_attack).

Plus, I'm not sure where is token generated? Cannot find where it is done.

Reply:
thank's
it's generated from the db, in the `add.iot.box` model, I'm working on it.

Reply:
I think we are talking about this token which was not originally made for this purpose.
https://github.com/odoo/enterprise/blob/master/iot/wizard/add_iot_box.py#L19

Reply:
Ok so if I understand correctly, this token is fixed in time? Does it change?
If not, what is preventing a leak?

Reply:
Thank's for the advice, to prevent from a leak, I changed the implementation:
- The DB still generates a token shared with the IoT Box on first connection,
- We still decorate routes with `@route.protect`,
- The db signs the (payload + timestamp) of its requests with the `hmac` method,
- The IoT Box route recorator will then retrieve the payload, sign it, and compare its signature with the provided one:
  - as the signature is time-based, the IoT Box will get the timestamp at the time it recieves the request, and sign with every timestamp in a 10 sec window
  - If one of those signatures is valid, it accepts the request.